url7507f7ef16_Neuro17_jsongkick_tgz-pJ8-jsongkick_AppJ8/src/http/SongkickConnector.java

```diff
@@ -0,0 +1,182 @@
+package http;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.lang.reflect.GenericArrayType;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.ArrayList;
+
+import org.apache.http.client.ClientProtocolException;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+import config.SongkickConfig;
+
+public abstract class SongkickConnector{
+	private static final Logger log = LogManager.getLogger(SongkickConnector.class);
+	private JsonObject jsonResponse;
+	protected Gson gson = new GsonBuilder().setPrettyPrinting().create();
+	protected URIBuilder uriBld;
+	protected URI uri;
+	protected Integer page;
+	protected int pages;
+	
+	protected Integer getPage() {
+		return page;
+	}
+
+	protected void setPage(int page) {
+		this.page = page;
+	}
+	
+	public boolean hasNextPage() {
+		return page < pages;
+	}
+
+	protected int getPages() {
+		return pages;
+	}
+
+	protected void setPages(int pages) {
+		this.pages = pages;
+	}
+	
+	protected URI getUri() {
+		return uri;
+	}
+
+	protected void setUri(URI uri) {
+		this.uri = uri;
+	}
+
+	protected URL url;
+	
+
+	/**
+	 * Builds a string representing response from songkick ready to be parse as JSON.
+	 * 
+	 * @param  response					HttpRespose received by HttpGet call.	
+	 * @return StringBuffer				String representing the response. Ready to parse as JSON.
+	 * @throws IllegalStateException
+	 * @throws IOException
+	 */
+	protected JsonObject parseResponseAsJson(InputStream response){
+		log.trace("Entering parseResponseAsJson");
+		BufferedReader rd = new BufferedReader(new InputStreamReader(response));
+		
+		JsonParser jsonParser = new JsonParser();
+		
+		StringBuffer result = new StringBuffer();
+		
+		String line = "";
+		
+		try {
+			while ((line = rd.readLine()) != null) {
+			    result.append(line);
+			}
+		} catch (IOException e) {
+			log.error(e.getMessage());
+		}
+		log.trace("Exiting parseResponseAsJson");
+		return jsonParser.parse(result.toString()).getAsJsonObject();
+	}
+	
+	/**
+	 * Performs HTTP get given a valid URI, returns JSON object representation of response.	
+	 * 
+	 * @param uri
+	 * @return JsonObject
+	 */
+	protected JsonObject executeRequest(URI uri){
+		//TODO - capire come accedere ai risultati che sono oltre la prima pagina. 
+		InputStream response;
+		
+		try {
+			url = uri.toURL();
+		} catch (MalformedURLException e1) {
+			
+		}
+				
+		try {
+			response = url.openStream();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *response*. Execution paths that do not contain closure statements were detected. Ensure *response* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).